File: lib/services/trakt/trakt_movies_api.dart
import 'package:dio/dio.dart';
import '../../core/config/trakt_config.dart';
import '../../models/trakt/trakt_movie.dart';

class TraktMoviesApi {
  final Dio _dio;
  
  TraktMoviesApi(this._dio);
  
  Future<List<TraktTrendingMovie>> getTrending({
    int page = 1,
    int limit = 10,
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/movies/trending',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return (response.data as List)
        .map((json) => TraktTrendingMovie.fromJson(json))
        .toList();
  }
  
  Future<List<TraktMovie>> getPopular({
    int page = 1,
    int limit = 10,
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/movies/popular',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (extended != null) 'extended': extended.value,
      },
    );
    
    // Debug logging
    print('Popular movies API response (first item): ${(response.data as List).isNotEmpty ? (response.data as List)[0] : "empty"}');
    
    return (response.data as List)
        .map((json) => TraktMovie.fromJson(json))
        .toList();
  }
  
  Future<List<TraktAnticipatedMovie>> getAnticipated({
    int page = 1,
    int limit = 10,
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/movies/anticipated',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return (response.data as List)
        .map((json) => TraktAnticipatedMovie.fromJson(json))
        .toList();
  }
  
  Future<List<TraktMovie>> getPlayed({
    int page = 1,
    int limit = 10,
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/movies/played',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return (response.data as List)
        .map((json) => TraktMovie.fromJson(json['movie'] ?? json))
        .toList();
  }
  
  Future<List<TraktMovie>> getWatched({
    String period = 'weekly',
    int page = 1,
    int limit = 10,
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/movies/watched/$period',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return (response.data as List)
        .map((json) => TraktMovie.fromJson(json['movie'] ?? json))
        .toList();
  }
  
  Future<List<TraktMovie>> getFavorited({
    String period = 'weekly',
    int page = 1,
    int limit = 10,
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/movies/favorited/$period',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return (response.data as List)
        .map((json) => TraktMovie.fromJson(json['movie'] ?? json))
        .toList();
  }
  
  Future<List<TraktMovie>> getCollected({
    String period = 'weekly',
    int page = 1,
    int limit = 10,
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/movies/collected/$period',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return (response.data as List)
        .map((json) => TraktMovie.fromJson(json['movie'] ?? json))
        .toList();
  }
  
  Future<TraktMovie> getSummary(
    String traktSlug, {
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/movies/$traktSlug',
      queryParameters: {
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return TraktMovie.fromJson(response.data);
  }
  
  Future<List<TraktMovie>> getRelated(
    String movieId, {
    int page = 1,
    int limit = 10,
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/movies/$movieId/related',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return (response.data as List)
        .map((json) => TraktMovie.fromJson(json))
        .toList();
  }
  
  Future<List<Map<String, dynamic>>> getMyMovies({
    String? startDate,
    int days = 7,
  }) async {
    final start = startDate ?? DateTime.now().toIso8601String().split('T')[0];
    
    final response = await _dio.get(
      '/calendars/my/movies/$start/$days',
      queryParameters: {
        'extended': 'full',
      },
    );
    
    return (response.data as List)
        .map((json) => json as Map<String, dynamic>)
        .toList();
  }
  
  Future<List<TraktMovie>> getRecentlyUpdatedMovies({
    int limit = 20,
  }) async {
    // Get date 7 days ago, rounded to the hour
    final startDate = DateTime.now().subtract(const Duration(days: 7));
    final formattedDate = '${startDate.year}-${startDate.month.toString().padLeft(2, '0')}-${startDate.day.toString().padLeft(2, '0')}T${startDate.hour.toString().padLeft(2, '0')}:00:00Z';
    
    final response = await _dio.get(
      '/movies/updates/$formattedDate',
      queryParameters: {
        'limit': limit,
        'extended': 'full',
      },
    );
    
    return (response.data as List)
        .map((json) => TraktMovie.fromJson(json['movie'] ?? json))
        .toList();
  }
}

File: lib/trakt/api/TraktMoviesApi.kt
package app.moviebase.trakt.api

import app.moviebase.trakt.TraktExtended
import app.moviebase.trakt.core.endPoint
import app.moviebase.trakt.core.parameterExtended
import app.moviebase.trakt.core.parameterLimit
import app.moviebase.trakt.core.parameterPage
import app.moviebase.trakt.model.TraktAnticipatedMovie
import app.moviebase.trakt.model.TraktMovie
import app.moviebase.trakt.model.TraktRating
import app.moviebase.trakt.model.TraktTrendingMovie
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.HttpRequestBuilder
import io.ktor.client.request.get

class TraktMoviesApi(
    private val client: HttpClient,
) {
    suspend fun getTrending(
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktTrendingMovie> =
        client
            .get {
                endPointMovies("trending")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getPopular(
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktMovie> =
        client
            .get {
                endPointMovies("popular")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getAnticipated(
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktAnticipatedMovie> =
        client
            .get {
                endPointMovies("anticipated")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getPlayed(
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktMovie> =
        client
            .get {
                endPointMovies("played")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getWatched(
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktMovie> =
        client
            .get {
                endPointMovies("watched")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getCollected(
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktMovie> =
        client
            .get {
                endPointMovies("collected")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getRelated(
        movieId: String,
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktMovie> =
        client
            .get {
                endPointMovie(movieId, "related")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getSummary(
        traktSlug: String,
        extended: TraktExtended? = null,
    ): TraktMovie =
        client
            .get {
                endPointMovies(traktSlug)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getRating(traktSlug: String): TraktRating =
        client
            .get {
                endPointMovie(traktSlug, "ratings")
            }.body()

    suspend fun getStats(traktSlug: String): TraktRating =
        client
            .get {
                endPointMovie(traktSlug, "stats")
            }.body()

    private fun HttpRequestBuilder.endPointMovie(
        movieId: String,
        vararg paths: String,
    ) {
        endPoint("movies", movieId, *paths)
    }

    private fun HttpRequestBuilder.endPointMovies(vararg paths: String) {
        endPoint("movies", *paths)
    }
}
